core/kazoo_web/src/kz_http.erl

```diff
@@ -1,260 +1,283 @@
-    Val = urldecode(ValAcc),
-
-    parse_query_string('key', R, <<>>, <<>>, RetAcc ++ [{Key, Val}]);
-
-parse_query_string('key', <<C, R/binary>>, KeyAcc, ValAcc, RetAcc) ->
-    parse_query_string('key', R, <<KeyAcc/binary, C>>, ValAcc, RetAcc);
-
-parse_query_string('val', <<C, R/binary>>, KeyAcc, ValAcc, RetAcc) ->
-    parse_query_string('val', R, KeyAcc, <<ValAcc/binary, C>>, RetAcc).
+-spec async_req(pid(), string()) -> kz_http_ret().
+-spec async_req(pid(), atom(), string()) -> kz_http_ret().
+-spec async_req(pid(), atom(), string(), wh_proplist()) -> kz_http_ret().
+-spec async_req(pid(), atom(), string(), wh_proplist(), http_body()) -> kz_http_ret().
+-spec async_req(pid(), atom(), string(), wh_proplist(), http_body(), wh_proplist()) -> kz_http_ret().
+async_req(Pid, Url) ->
+    async_req(Pid, 'get', Url, [], [], []).
+async_req(Pid, Method, Url) ->
+    async_req(Pid, Method, Url, [], [], []).
+async_req(Pid, Method, Url, Headers) ->
+    async_req(Pid, Method, Url, Headers, []).
+async_req(Pid, Method, Url, Headers, Body) ->
+    async_req(Pid, Method, Url, Headers, Body, []).
+async_req(Pid, Method, Url, Hdrs, Body, Opts) ->
+    {Headers, Options} = maybe_basic_auth(Hdrs, Opts),
+    Request = build_request(Method, Url, Headers, Body),
+    execute_request(Method, Request, [{receiver, Pid}, {sync, false}, {stream, self} | Options]).
 
 %%--------------------------------------------------------------------
-%% @doc Splits a URL into scheme, location, path, query, and fragment parts
+%% @private
+%% @doc Send request using httpc and handle its response
+%% @end
 %%--------------------------------------------------------------------
--spec urlsplit(binary()) -> {binary(), binary(), binary(), binary(), binary()}.
-urlsplit(Source) ->
-    {Scheme, Url1}      = urlsplit_s(Source),
-    {Location, Url2}    = urlsplit_l(Url1),
-    {Path, Query, Frag} = urlsplit_p(Url2, <<>>),
-
-    {Scheme, Location, Path, Query, Frag}.
+-spec execute_request(atom(), tuple(), wh_proplist()) -> kz_http_ret().
+execute_request(Method, Request, Opts) ->
+    HTTPOptions = get_options(?HTTP_OPTIONS, Opts),
+    Opts1 = get_options(?OPTIONS, Opts),
+    Options = case props:get_value('body_format', Opts1) of
+                'undefined' -> [{'body_format', 'binary'} | Opts1];
+                _ -> Opts1
+              end,
+    handle_response(catch httpc:request(Method, Request, HTTPOptions, Options)).
 
 %%--------------------------------------------------------------------
-%% @private
-%% @doc Splits out the scheme portion of the URL (if present)
+%% @public
+%% @doc
+%% Response to caller in a proper manner
+%% @end
 %%--------------------------------------------------------------------
--spec urlsplit_s(binary()) -> {binary(), binary()}.
-urlsplit_s(Source) ->
-    case urlsplit_s(Source, <<>>) of
-        'no_scheme' -> {<<>>, Source};
-        ValidScheme -> ValidScheme
+-spec handle_response(httpc_ret()) -> kz_http_ret().
+handle_response(Resp) ->
```

Comment:
Just use the function clauses to handle the Resp varieties instead of using a case here. Then there's no need for `handle_good_response/1` either